### PR TITLE
clone3() and changes in sigaltstack to work with glibc 2.34

### PR DESCRIPTION
**Pull request by gnode1**
Fedora 35 brings in glibc-2.34, which uses clone3() and dynamic stack size. This makes changes to support theses.

Tested manually running tests on Fedora 35. Separate clone3() test isn't implemented s musl has no wrapper for it.

There are still some other failures on Fedora 35 static glibc tests so this is not fully done yet.

### CODE REVIEW

File: payloads/java/Makefile [hunk 1]
```diff
@@ -82,8 +82,8 @@ fromsrc: .check ${JDK_VERSION} jtreg ${KM_BIN} ${KM_LDSO}
 	fi
 	@# -fcommon gcc flag is needed for gcc10, see https://bugs.openjdk.java.net/browse/JDK-8235903
 	cd ${JDK_VERSION} && bash configure \
-		--disable-warnings-as-errors --with-native-debug-symbols=none \
-		--with-extra-cflags='-fcommon'  \
+		--disable-warnings-as-errors --with-native-debug-symbols=internal \
```
Comment (johnmuth81) [inline payloads/java/Makefile:85]: Yeah, I see what you mean. I think this was just source code entropy in action. Good catch.

**Review by johnmuth81 — COMMENTED** [review]
I think the java thing was just code drift. Good catch.

Added some other comments.

File: km/km.h [hunk 2]
```diff
@@ -31,6 +31,7 @@
 #include <sys/param.h>
 #include <sys/types.h>
 #include <linux/kvm.h>
+#include <linux/sched.h>
```
[the remark this answers is not in the document]
Comment (gnode1) [reply km/km.h:34]: right

File: km/km_cpu_init.c [hunk 3]
```diff
@@ -667,10 +670,12 @@ void km_machine_setup(km_machine_init_params_t* params)
             __get_cpuid(entry->function, &entry->eax, &entry->ebx, &entry->ecx, &entry->edx);
             break;
          case 0x0:
+            __get_cpuid(entry->function, &entry->eax, &entry->ebx, &entry->ecx, &entry->edx);
+
             km_infox(KM_TRACE_KVM, "Setting VendorId to '%s'", cpu_vendor_id);
```
[the remark this answers is not in the document]
Comment (gnode1) [reply km/km_cpu_init.c:675]: I'm still working on this one.

**Comment by gnode1** [comment]
I think this is ready to be merged. Our tests work fine. The node tests, even native node tests, require changes in the system SSL settings, but even then node v12 still has one test that fails. I'm skipping this test or now. I think the right answer there is to move to more current version of node, in a separate PR

File: km/km_cpu_init.c [hunk 4]
```diff
@@ -667,10 +671,12 @@ void km_machine_setup(km_machine_init_params_t* params)
...
-            km_infox(KM_TRACE_KVM, "Setting VendorId to '%s'", cpu_vendor_id);
-            memcpy(&entry->ebx, cpu_vendor_id, sizeof(u_int32_t));
-            memcpy(&entry->edx, cpu_vendor_id + sizeof(u_int32_t), sizeof(u_int32_t));
-            memcpy(&entry->ecx, cpu_vendor_id + 2 * sizeof(u_int32_t), sizeof(u_int32_t));
+            if (set_cpu_vendor_id != 0) {
+               km_infox(KM_TRACE_KVM, "Setting VendorId to '%s'", cpu_vendor_id);
+               memcpy(&entry->ebx, cpu_vendor_id, sizeof(u_int32_t));
+               memcpy(&entry->edx, cpu_vendor_id + sizeof(u_int32_t), sizeof(u_int32_t));
+               memcpy(&entry->ecx, cpu_vendor_id + 2 * sizeof(u_int32_t), sizeof(u_int32_t));
+            }
```
Comment (sv641) [inline km/km_cpu_init.c:679]: how does it work when set_cpu_vendor_id is not set.

File: km/km_hcalls.c [hunk 5]
```diff
@@ -1432,6 +1432,23 @@ static km_hc_ret_t clone_hcall(void* vcpu, int hc, km_hc_args_t* arg)
    return HC_CONTINUE;
 }
 
+static km_hc_ret_t clone3_hcall(void* vcpu, int hc, km_hc_args_t* arg)
+{
+   // long syscall(SYS_clone3, struct clone_args *cl_args, size_t size);
+   struct clone_args* cl_args = km_gva_to_kma(arg->arg1);
+
+   if (cl_args == NULL) {
```
Comment (sv641) [inline km/km_hcalls.c:1440]: we generally do end of structure check to be in bounds.

File: km/km_cpu_init.c (hunk 4 above)
Comment (gnode1) [reply km/km_cpu_init.c:679]: `entry` is an element from `machine.entries` array which is set by `KVM_GET_SUPPORTED_CPUID` ioctl a little above, so it is already set up. I'm not sure why 0x80000002, 0x80000003 and 0x80000004 are treated special, haven't really dug into it

File: km/km_hcalls.c (hunk 5 above)
Comment (gnode1) [reply km/km_hcalls.c:1440]: Not sure this is actually correct. I see a lot of places where we only check the address. Can you point at examples?

Comment (sv641) [reply km/km_hcalls.c:1440]: uname hcall.

**Review by sv641 — APPROVED** [review]

**Pull request merged**